### DISCUSSION TITLE
Define a default for a Generic class

### DISCUSSION THREAD

**Discussion by Kaniee** [microsoft/pyright#5555] [Q&A]
Hi there, I struggle to phrase my question well, so let me directly share the MWE code:
```python
from typing import TypeVar, Generic, Callable

T = TypeVar("T", bound=float)

class FloatCaster(Generic[T]):
    get_one: Callable[[str], T] = float

    @classmethod
    def get_three(cls, input: str) -> list[T]:
        return [cls.get_one(input) for _ in range(3)]

class IntCaster(FloatCaster[int]):
    get_one = int

three_floats = FloatCaster.get_three("7")  # This should be of type list[float]
three_ints = IntCaster.get_three("7")  # This should be of type list[int]
```

I'd like to set up the parent class (`FloatCaster`) in such a way, that a child class can overwrite the `get_one` method and the new return type of the `get_one` method would be reflected in the return type of `get_three`.

pyright and mypy are complaining about `get_one: Callable[[str], T] = float`.
I feels like I should define some sort of default type for `T` but I don't know how or whether this is the right direction.
I'm also happy to get rid of the Generic, if that is possible. I just need the list return type of `get_three` to be inferred by the return type of `get_one`.

I feel like there is a obvious solution to this just around the corner but I can't see it.

I appreciate your help.

**Comment by erictraut** [comment]
Looks like a job for an abstract base class.

```python
from abc import ABC, abstractmethod
from typing import TypeVar, Generic

T = TypeVar("T")

class Caster(Generic[T], ABC):
    @abstractmethod
    @classmethod
    def get_one(cls, input: str) -> T:
        ...

    @classmethod
    def get_three(cls, input: str) -> list[T]:
        return [cls.get_one(input) for _ in range(3)]

class FloatCaster(Caster[float]):
    @classmethod
    def get_one(cls, input: str) -> float:
        return float(input)

class IntCaster(Caster[int]):
    @classmethod
    def get_one(cls, input: str) -> int:
        return int(input)
```

**Reply by Kaniee** [reply]
Thanks for the suggestion, that would solve the problem I stated.

But I forgot to mention that I'm looking for a solution where `IntCaster` is still a child class of `FloatCaster` since it is a specialization (i.e. int is a acceptable return type for float).
And I was hoping to not add types/complexity to satisfy typing.

Do you know of a solution that would allow `IntCaster` to inherit from `FloatCaster`?

**Reply by erictraut** [reply]
```python
from abc import ABC, abstractmethod
from typing import TypeVar, Generic

T = TypeVar("T", covariant=True)

class Caster(Generic[T], ABC):
    @classmethod
    @abstractmethod
    def get_one(cls, input: str) -> T:
        ...

    @classmethod
    def get_three(cls, input: str) -> list[T]:
        return [cls.get_one(input) for _ in range(3)]

class FloatCaster(Caster[float]):
    @classmethod
    def get_one(cls, input: str) -> float:
        return float(input)

class IntCaster(FloatCaster):
    @classmethod
    def get_one(cls, input: str) -> int:
        return int(input)

print(IntCaster.get_three("2"))
print(FloatCaster.get_three("2"))
```

**Reply by Kaniee** [reply]
Thanks for the response.
Unfortunately with this approach `IntCaster.get_three("2")` is of type `list[float]` while I'm looking for a solution where this is `list[int]`.

I tried to inform the `IntCaster` about the generic type but with no effect:
```python
class IntCaster(FloatCaster, Caster[int]):
    @classmethod
    def get_one(cls, input: str) -> int:
        return int(input)

result = IntCaster.get_three("2")  # Unfortunately list[float]
```
